pkg/services/ngalert/state/manager.go

```diff
@@ -177,26 +177,22 @@ func (st *Manager) ProcessEvalResults(ctx context.Context, evaluatedAt time.Time
 		s := st.setNextState(ctx, alertRule, result, extraLabels, logger)
 		states = append(states, s)
 	}
-	resolvedStates := st.staleResultsHandler(ctx, evaluatedAt, alertRule, logger)
+	staleStates := st.deleteStaleStatesFromCache(ctx, logger, evaluatedAt, alertRule)
+	st.deleteAlertStates(ctx, logger, staleStates)
 
 	st.saveAlertStates(ctx, logger, states...)
 
-	changedStates := make([]StateTransition, 0, len(states))
-	for _, s := range states {
-		if s.changed() {
-			changedStates = append(changedStates, s)
-		}
-	}
-
-	if st.historian != nil {
-		st.historian.RecordStates(ctx, alertRule, changedStates)
-	}
+	st.logStateTransitions(ctx, alertRule, states, staleStates)
 
-	deltas := append(states, resolvedStates...)
 	nextStates := make([]*State, 0, len(states))
-	for _, s := range deltas {
+	for _, s := range states {
 		nextStates = append(nextStates, s.State)
 	}
+	for _, s := range staleStates {
+		if s.PreviousState == eval.Alerting {
```

Comment:
We can eliminate this filtering if the deletion handler only returns resolved states to begin with.

Reply:
This will move database operations back to deleteStaleStates method. The idea of this PR is to consolidate all database operations. Filtering will not make a sensible impact because stale states usually are pretty rare